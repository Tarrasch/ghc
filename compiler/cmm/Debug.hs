{-# LANGUAGE GADTs #-}
-----------------------------------------------------------------------------
--
-- Debugging data
--
-- Association of debug data on the Cmm level, with methods to encode it in
-- event log format for later inclusion in profiling event logs.
--
-----------------------------------------------------------------------------

module Debug (

  DebugBlock(..), dblIsEntry,
  UnwindTable, UnwindExpr(..),
  cmmDebugGen,
  cmmDebugLink,
  debugToMap

  ) where

import BlockId         ( blockLbl, BlockEnv )
import CLabel
import Cmm
import CmmUtils
import CoreSyn
import FastString      ( nilFS, mkFastString )
import Module
import Outputable
import PprCore         ()
import PprCmmExpr      ( pprExpr )
import SrcLoc

import Compiler.Hoopl

import Data.Maybe
import Data.List     ( find, minimumBy )
import Data.Ord      ( comparing )
import qualified Data.Map as Map

-- | Debug information about a block of code. Context is encoded through nesting.
data DebugBlock =
  DebugBlock
  { dblProcedure  :: !Label              -- ^ Entry label of containing producedure
  , dblLabel      :: !Label              -- ^ Hoopl label
  , dblCLabel     :: !CLabel             -- ^ Output label
  , dblHasInfoTbl :: !Bool               -- ^ Has an info table?
  , dblTicks      :: ![RawTickish]       -- ^ Ticks defined in this block
  , dblSourceTick :: !(Maybe RawTickish) -- ^ Best source tick covering this block
  , dblPosition   :: !(Maybe Int)        -- ^ Output position relative to other blocks in proc.
                                         --   @Nothing@ means the block has been optimized out.
  , dblUnwind     :: !UnwindTable        -- ^ Unwind information
  , dblBlocks     :: ![DebugBlock]       -- ^ Nested blocks
  }

dblIsEntry :: DebugBlock -> Bool
dblIsEntry blk = dblProcedure blk == dblLabel blk

instance Outputable DebugBlock where
  ppr blk = (if dblHasInfoTbl blk
             then if dblProcedure blk == dblLabel blk
                  then text "proc "
                  else text "pp-blk "
             else text "blk ") <>
            ppr (dblLabel blk) <+> parens (ppr (dblCLabel blk)) <+>
            (maybe empty ppr (dblSourceTick blk)) <+>
            (maybe empty ppr (find isCore (dblTicks blk))) <+>
            (maybe (text "removed") ((text "pos " <>) . ppr) (dblPosition blk)) <+>
            pprUwMap (dblUnwind blk) $$
            (if null (dblBlocks blk) then empty else ppr (dblBlocks blk))
    where isCore CoreNote{} = True
          isCore _other     = False
          pprUw (g, expr) = ppr g <> char '=' <> ppr expr
          pprUwMap = braces . hsep . punctuate comma . map pprUw . Map.toList

-- | Extract debug data from a group of procedures
cmmDebugGen :: ModLocation -> RawCmmGroup -> [DebugBlock]
cmmDebugGen loc decls =
  let isProc CmmProc{} = True
      isProc _other    = False
      procs = filter isProc decls

      -- Map of blocks to the procedure they were declared in
      procMap :: LabelMap RawCmmDecl
      procMap = mapFromList $ concatMap procMappings procs
      procMappings p = [ (entryLabel b, p) | CmmProc _ _ _ g <- [p], b <- toBlockList g]

      -- Info tables
      infos = mapUnions $ map (\(CmmProc i _ _ _) -> i) procs
      hasInfoTable l = l `mapMember` infos

      -- Combined block and context maps
      blockMap :: BlockEnv CmmBlock
      blockMap = mapUnions $ map (\(CmmProc _ _ _ g) -> toBlockMap g) procs
      contextMap :: BlockEnv Label
      contextMap = mapUnions $ map (\(CmmProc _ _ _ g) -> getContextMap g) procs

      topBlocks :: BlockEnv CmmBlock
      topBlocks = blockMap `mapDifference` mapMap (const undefined) contextMap -- ugly

      -- Helper: Best source ticks from a list
      --
      -- This is somewhat arbitrary -- we select the first source
      -- span, while preferring source ticks from the same source
      -- file.  Furthermore, dumps take priority (if we generated one,
      -- we probably want debug information to refer to it).
      bestSrcTick = minimumBy (comparing rangeRating)
      rangeRating (SourceNote span _)
        | isDumpSrcSpan span           = 0
        | srcSpanFile span == thisFile = 1
        | otherwise                    = 2 :: Int
      rangeRating _non_source_note     = error "rangeRating"
      thisFile = maybe nilFS mkFastString $ ml_hs_file loc
      isSourceTick SourceNote {} = True
      isSourceTick _             = False

      mkDbgBlock cstick cproc lbl block
        -- Make sure we never have a block in the context tree that's
        -- not a direct or indirect child of its procedure entry
        -- block. This might fail for two reasons:
        -- * cproc is Nothing, in which case this must be a naked
        --   block without context. This is probably CodeGen
        --   forgetting to generate a suitable CmmContext directive
        --   for a new block.
        -- * Otherwise, the CmmContext led us to an internal block
        --   from another procedure. While this is not fatal at this
        --   point, it's just no good idea to refer to internal labels
        --   from another procedure.
        | procEntry /= lbl && Just procEntry /= cproc
        = pprPanic "cmmDebugGen: Block has missing or inconsistent procedure context! "
          (ppr lbl <> text ": expected " <> ppr procEntry <> text ", got " <> ppr cproc)
        | otherwise
        = DebugBlock { dblProcedure    = procEntry
                     , dblLabel        = lbl
                     , dblCLabel       =
                         if g_entry graph == lbl
                         then case mapLookup lbl infos of
                                Just (Statics il _) -> il
                                _other              -> el
                         else blockLbl lbl
                     , dblHasInfoTbl   = hasInfoTable lbl
                     , dblTicks        = ticks
                     , dblPosition     = Nothing -- updated later by cmmDebugLink
                     , dblUnwind       = extractUnwind block
                     , dblSourceTick   = stick
                     , dblBlocks       = mapMaybe mkByLabel (blockContexts block)
                     }
        where mkByLabel l = fmap (mkDbgBlock stick (Just procEntry) l) (mapLookup l blockMap)
              Just (CmmProc infos el _ graph) = mapLookup lbl procMap
              procEntry = g_entry graph
              -- A source tick scopes over nested blocks as long as
              -- they don't have source ticks of their own, which
              -- otherwise take priority.
              ticks  = blockTicks block
              sticks = filter isSourceTick ticks
              stick  = if null sticks then cstick else Just $! bestSrcTick sticks

  in map (uncurry (mkDbgBlock Nothing Nothing)) $ mapToList topBlocks

-- | Sets position fields in the debug block tree according to native
-- generated code.
cmmDebugLink :: (i -> Bool) -> GenCmmGroup d g (ListGraph i)
                -> [DebugBlock] -> [DebugBlock]
cmmDebugLink isMeta nats blocks =
  let -- Find order in which procedures will be generated by the
      -- back-end (important for DWARF).
      --
      -- Note that we might encounter blocks that are missing or only
      -- consist of meta instructions -- we will declare them missing,
      -- which will prevent debug data generation without messing up
      -- the block hierarchy.
      getBlocks (CmmProc _ _ _ (ListGraph bs)) = bs
      getBlocks _other                         = []
      allMeta (BasicBlock _ instrs) = all isMeta instrs
      blockPosition :: LabelMap Int
      blockPosition = mapFromList $ flip zip [0..] $ map blockId $ filter (not . allMeta) $
                      concatMap getBlocks nats

      link block = block { dblPosition = mapLookup (dblLabel block) blockPosition
                         , dblBlocks   = map link (dblBlocks block)
                         }
   in map link blocks

-- | Converts debug blocks into a label map for easier lookups
debugToMap :: [DebugBlock] -> LabelMap DebugBlock
debugToMap = mapUnions . map go
   where go b = mapInsert (dblLabel b) b $ mapUnions $ map go (dblBlocks b)

-- | Maps registers to expressions that yield their "old" values
-- further up the stack. Most interesting for the stack pointer Sp,
-- but might be useful to document saved registers, too.
type UnwindTable = Map.Map GlobalReg UnwindExpr

-- | Expressions, used for unwind information
data UnwindExpr = UwConst Int                   -- ^ literal value
                | UwReg GlobalReg Int           -- ^ register plus offset
                | UwDeref UnwindExpr            -- ^ pointer dereferencing
                | UwPlus UnwindExpr UnwindExpr
                | UwMinus UnwindExpr UnwindExpr
                | UwTimes UnwindExpr UnwindExpr
                deriving (Eq)

instance Outputable UnwindExpr where
  pprPrec _ (UwConst i)     = ppr i
  pprPrec _ (UwReg g 0)     = ppr g
  pprPrec p (UwReg g x)     = pprPrec p (UwPlus (UwReg g 0) (UwConst x))
  pprPrec _ (UwDeref e)     = char '*' <> pprPrec 3 e
  pprPrec p (UwPlus e0 e1)  | p <= 0
                            = pprPrec 0 e0 <> char '+' <> pprPrec 0 e1
  pprPrec p (UwMinus e0 e1) | p <= 0
                            = pprPrec 1 e0 <> char '-' <> pprPrec 1 e1
  pprPrec p (UwTimes e0 e1) | p <= 1
                            = pprPrec 2 e0 <> char '*' <> pprPrec 2 e1
  pprPrec _ other           = parens (pprPrec 0 other)

extractUnwind :: CmmBlock -> UnwindTable
extractUnwind b = go $ blockToList mid
  where (_, mid, _) = blockSplit b
        go :: [CmmNode O O] -> UnwindTable
        go []       = Map.empty
        go (x : xs) = case x of
          CmmUnwind g so -> Map.insert g (toUnwindExpr so) $! go xs
          CmmTick {}     -> go xs
          CmmContext {}  -> go xs
          _other         -> Map.empty -- TODO: Unwind statements after actual instructions

-- | Conversion of Cmm expressions to unwind expressions. We check for
-- unsupported operator usages and simplify the expression as far as
-- possible.
toUnwindExpr :: CmmExpr -> UnwindExpr
toUnwindExpr (CmmLit (CmmInt i _))       = UwConst (fromIntegral i)
toUnwindExpr (CmmRegOff (CmmGlobal g) i) = UwReg g i
toUnwindExpr (CmmReg (CmmGlobal g))      = UwReg g 0
toUnwindExpr (CmmLoad e _)               = UwDeref (toUnwindExpr e)
toUnwindExpr e@(CmmMachOp op [e1, e2])   =
  case (op, toUnwindExpr e1, toUnwindExpr e2) of
    (MO_Add{}, UwReg r x, UwConst y) -> UwReg r (x + y)
    (MO_Sub{}, UwReg r x, UwConst y) -> UwReg r (x - y)
    (MO_Add{}, UwConst x, UwReg r y) -> UwReg r (x + y)
    (MO_Add{}, UwConst x, UwConst y) -> UwConst (x + y)
    (MO_Sub{}, UwConst x, UwConst y) -> UwConst (x - y)
    (MO_Mul{}, UwConst x, UwConst y) -> UwConst (x * y)
    (MO_Add{}, u1,        u2       ) -> UwPlus u1 u2
    (MO_Sub{}, u1,        u2       ) -> UwMinus u1 u2
    (MO_Mul{}, u1,        u2       ) -> UwTimes u1 u2
    _otherwise                       -> pprPanic "Unsupported operator in unwind expression!" (pprExpr e)
toUnwindExpr e
  = pprPanic "Unsupported unwind expression!" (ppr e)

