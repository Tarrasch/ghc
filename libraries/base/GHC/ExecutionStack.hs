-----------------------------------------------------------------------------
-- |
-- Module      :  GHC.ExecutionStack
-- Copyright   :  (c) The University of Glasgow 2013-2015
-- License     :  see libraries/base/LICENSE
--
-- Maintainer  :  cvs-ghc@haskell.org
-- Stability   :  internal
-- Portability :  non-portable (GHC Extensions)
--
-- This is a module for efficient stack traces. You can fully use module
-- without any rts-flags to be set.  This stack trace implementation is
-- considered low overhead.
--
-- @
-- import GHC.ExecutionStack
--
-- myFunction :: IO ()
-- myFunction = do
--      stackFrames <- currentStackFrames
--      putStrLn $ displayStackFrames stackFrames
-- @
--
-- /Since: 4.11.0.0/
-----------------------------------------------------------------------------

module GHC.ExecutionStack (
    currentStackFrames
  , displayStackFrames
  ) where

import GHC.ExecutionStack.Internal (currentStackFrames, displayStackFrames)
