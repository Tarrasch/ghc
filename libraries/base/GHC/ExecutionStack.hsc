-----------------------------------------------------------------------------
-- |
-- Module      :  GHC.ExecutionStack
-- Copyright   :  (c) The University of Glasgow 2013-2015
-- License     :  see libraries/base/LICENSE
--
-- Maintainer  :  cvs-ghc@haskell.org
-- Stability   :  internal
-- Portability :  non-portable (GHC Extensions)
--
-- This is a module for the efficient but inaccurate stack traces. If you
-- can take a factor 2 of performance penalty, you should consider using
-- "GHC.Stack" as the stack traces will be more accurate and detailed.
--
-- @
-- myFunction :: IO ()
-- myFunction = do
--      stack <- currentExecutionStack 100
--      printExecutionStack stack
-- @
--
-- An 'ExecutionStack' is a data wrapper around 'ByteArray#'. The
-- Array is a reified stack. Each element is a code address. For most
-- frames this will be the return address for the stack frame, but
-- for an update frame the address is the entry code of the thunk (if
-- available)
--
-- /Since: 4.11.0.0/
-----------------------------------------------------------------------------

{-# LANGUAGE UnboxedTuples, MagicHash, RecordWildCards #-}
module GHC.ExecutionStack (
  -- * Simple interface
    printCurrentExecutionStack
  , currentExecutionStack
  , printExecutionStack
  -- * Intermediate interface (May change)
  , ExecutionStack ()
  , currentExecutionStackLimit
  , getStackFrames
  , StackFrame(..)
  -- * Complicated interface (May change)
  -- ** Managed loading/unloading
  , inCodemap
  , codemapIncRef
  , codemapDecRef
  , codemapTryUnload
  , codemapIsLoaded
  -- ** Looking inside the `ExecutionStack` value
  , stackSize
  , stackIndex
  , stackIndices
  -- ** Looking up instructions
  , getStackFramesNoSync
  , getStackFrame
  , getStackFrameNoSync
  ) where

import GHC.IO (IO(..))
import GHC.Base (Int(I##))
import GHC.Ptr (Ptr(Ptr))
import GHC.Prim (ByteArray##, sizeofByteArray##, indexAddrArray##)
import GHC.Prim (reifyStack##)
import Foreign.C.String (peekCString, CString)
import Foreign.Ptr (nullPtr)
import Foreign.Storable (Storable(..))
import Foreign.Marshal (alloca)
import Text.Printf (printf)
import Control.Exception.Base (bracket_)

#include "Rts.h"

data ExecutionStack = ExecutionStack
    { unExecutionStack ::  ByteArray##
    }

-- | The number of functions on your stack
stackSize :: ExecutionStack -> Int
stackSize stack =
    I## (sizeofByteArray## (unExecutionStack stack)) `div` (#const SIZEOF_HSPTR)

stackIndex :: ExecutionStack -> Int -> Ptr Instruction
stackIndex (ExecutionStack ba##) (I## i##) = Ptr (indexAddrArray## ba## i##)

stackIndices :: ExecutionStack -> [Ptr Instruction]
stackIndices stack = map (stackIndex stack) [0..(stackSize stack)-1]

data StackFrame = StackFrame
    { unitName      :: !String -- ^ From symbol table
    , procedureName :: !String -- ^ From symbol table
    } deriving Show

-- | Like 'show', without @unlines@
prepareStackFrame :: StackFrame -> [String]
prepareStackFrame su =
        [procedureName su]
    --  Note: We intend to have multiple lines per frame once we have dwarf


-- | Pretty print the execution stack to stdout
printExecutionStack :: ExecutionStack -> IO ()
printExecutionStack stack = do
    frames <- getStackFrames stack
    putStrLn $ displayFramesWithHeader frames

displayFramesWithHeader :: [StackFrame] -> String
displayFramesWithHeader frames =
    "Stack trace (Haskell):\n" ++
    concatMap (uncurry displayFrame) ([0..] `zip` frames)

-- | How one StackFrame is displayed in one stack trace
displayFrame :: Int -> StackFrame -> String
displayFrame ix frame = unlines $ zipWith ($) formatters strings
      where formatters = (printf "%4u: %s" (ix :: Int)) : repeat ("      " ++)
            strings    = prepareStackFrame frame

-- We use these three empty data declarations for type-safety
data CodemapUnit
data CodemapProc
data Instruction

peekCodemapUnitName :: Ptr CodemapUnit -> IO CString
peekCodemapUnitName ptr = #{peek struct CodemapUnit_, name } ptr

peekCodemapProcName :: Ptr CodemapProc -> IO CString
peekCodemapProcName ptr = #{peek struct CodemapProc_, name } ptr

-- | Print the current execution stack. The result printed to stdout should
-- look something like this:
--
-- @
-- Loading debug data...
-- Stack trace:
--    1: reifyStack.(...) (at libraries/base/GHC/ExecutionStack.hs:87:54-87:67)
--    2: bindIO (at libraries/base/GHC/Base.lhs:609:61-609:77)
--    3: stg_catch_frame_ret (at rts/Exception.cmm:365:1-367:2)
-- @
--
-- Even though the stack trace is somewhat cryptic, it could be as useful as a
-- tiny lantern in complete darkness.
printCurrentExecutionStack :: IO ()
printCurrentExecutionStack = currentExecutionStack >>= printExecutionStack

-- | Reify the stack. This is the only way to get an ExecutionStack value.
currentExecutionStack :: IO (ExecutionStack)
currentExecutionStack = currentExecutionStackLimit (maxBound :: Int)


currentExecutionStackLimit :: Int -> IO (ExecutionStack)
currentExecutionStackLimit (I## i##) =
    IO (\s -> let (## new_s, byteArray## ##) = reifyStack## i## s
                  ba = ExecutionStack byteArray##
              in (## new_s, ba ##) )

-- | Tell the codemap module that you want to use codemap. Synchronized.
foreign import ccall "Codemap.h codemap_inc_ref" codemapIncRef :: IO ()

-- | Tell the codemap module that you are done using codemap. Synchronized.
foreign import ccall "Codemap.h codemap_dec_ref" codemapDecRef :: IO ()

-- | Ask the codemap module if it can unload and free up memory. It will not be
-- able to if the module is in use or if data is not loaded. Synchronized.
--
-- Returns True if codemap data was unloaded.
foreign import ccall "Codemap.h codemap_try_unload" codemapTryUnload :: IO Bool

foreign import ccall "Codemap.h codemap_is_loaded" codemapIsLoaded :: IO Bool

-- | Lookup an instruction pointer
--
-- Codemap module must be loaded to use this!
foreign import ccall "Codemap.h codemap_lookup_ip"
    codemapLookupIp ::
       Ptr Instruction -- ^ Code address you want information about
    -> Ptr (Ptr CodemapProc) -- ^ Out: CodemapProc Pointer Pointer
    -> Ptr (Ptr CodemapUnit) -- ^ Out: CodemapUnit Pointer Pointer
    -> IO ()

inCodemap :: IO a -> IO a
inCodemap = bracket_ codemapIncRef codemapDecRef

getStackFrame ::
       Ptr Instruction -- ^ Instruction Pointer
    -> IO StackFrame -- ^ Result
getStackFrame ip =
    inCodemap $ getStackFrameNoSync ip

getStackFrameNoSync ::
       Ptr Instruction -- ^ Instruction Pointer
    -> IO StackFrame -- ^ Result
getStackFrameNoSync ip = do
    alloca $ \ppCodemapProc -> do
      poke ppCodemapProc nullPtr
      alloca $ \ppCodemapUnit -> do
          codemapLookupIp ip
                        ppCodemapProc
                        ppCodemapUnit
          pCodemapProc <- peek ppCodemapProc
          pCodemapUnit <- peek ppCodemapUnit
          unitName <- stringPeekWith peekCodemapUnitName pCodemapUnit
          procedureName <- stringPeekWith peekCodemapProcName pCodemapProc
          return StackFrame{..}

-- Note: if you grepped your way to the string "<Data not found>,
-- you probably forgot to compile that module with the `-g` flag to ghc.
stringPeekWith :: (Ptr a -> IO CString) -> Ptr a -> IO String
stringPeekWith _peeker ptr | ptr == nullPtr = return "<Data not found>"
stringPeekWith peeker ptr  | otherwise      = peeker ptr >>= peekCString

getStackFrames :: ExecutionStack -> IO [StackFrame]
getStackFrames stack = inCodemap $ getStackFramesNoSync stack

getStackFramesNoSync :: ExecutionStack -> IO [StackFrame]
getStackFramesNoSync = mapM getStackFrameNoSync . stackIndices
