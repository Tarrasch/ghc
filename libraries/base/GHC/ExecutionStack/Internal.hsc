-----------------------------------------------------------------------------
-- |
-- Module      :  GHC.ExecutionStack.Internal
-- Copyright   :  (c) The University of Glasgow 2013-2015
-- License     :  see libraries/base/LICENSE
--
-- Maintainer  :  cvs-ghc@haskell.org
-- Stability   :  internal
-- Portability :  non-portable (GHC Extensions)
--
-- Internals of `GHC.ExecustionStack` module
--
-- /Since: 4.11.0.0/
-----------------------------------------------------------------------------

{-# LANGUAGE UnboxedTuples, MagicHash, RecordWildCards #-}
module GHC.ExecutionStack.Internal (
  -- * Internal
    currentStackFrames
  , displayStackFrames
  , currentExecutionStack
  , codemapTryUnload
  , codemapIsLoaded
  , withCodemap
  , getStackFrames
  , getStackFramesNoSync
  ) where

import GHC.IO (IO(..))
import GHC.Base (Int(I##))
import GHC.Ptr (Ptr(Ptr))
import GHC.Prim (ByteArray##, sizeofByteArray##, indexAddrArray##)
import GHC.Prim (reifyStack##)
import Foreign.C.String (peekCString, CString)
import Foreign.Ptr (nullPtr)
import Foreign.Storable (Storable(..))
import Foreign.Marshal (alloca)
import Text.Printf (printf)
import Control.Exception.Base (bracket_)

#include "Rts.h"

-- | An 'ExecutionStack' represents a reified stack at some moment and each
-- element is a code address.
--
data ExecutionStack = ExecutionStack
    { unExecutionStack ::  ByteArray##
    }

-- | The number of functions on your stack
stackSize :: ExecutionStack -> Int
stackSize stack =
    I## (sizeofByteArray## (unExecutionStack stack)) `div` (#const SIZEOF_HSPTR)

stackIndex :: ExecutionStack -> Int -> Ptr Instruction
stackIndex (ExecutionStack ba##) (I## i##) = Ptr (indexAddrArray## ba## i##)

stackIndices :: ExecutionStack -> [Ptr Instruction]
stackIndices stack = map (stackIndex stack) [0..(stackSize stack)-1]

data StackFrame = StackFrame
    { unitName      :: !String -- ^ From symbol table
    , procedureName :: !String -- ^ From symbol table
    -- Once we have some more interesting data here, like with dwarf data, it
    -- might make sense to export this function and getStackFrames
    } deriving Show

-- | Like 'show', without @unlines@
prepareStackFrame :: StackFrame -> [String]
prepareStackFrame su =
        [procedureName su]
    --  Note: We intend to have multiple lines per frame once we have dwarf

-- | Pretty print a stack:
--
-- @
-- Stack trace (Haskell):
--    0: base_GHCziExecutionStackziInternal_currentStackFrames1_info
--    1: base_GHCziBase_bindIO1_info
--    2: base_GHCziIO_unsafeDupablePerformIO_info
--    3: sPd_info
--    4: base_GHCziShow_zdfShowZLZRzuzdcshow_info
--    5: sPe_info
--    6: base_GHCziBase_eqString_info
--    7: Main_dontInlineMe_info
--    8: stg_bh_upd_frame_info
--    9: base_GHCziShow_showLitString_info
--   10: s2YV_info
--   11: base_GHCziIOziHandleziText_zdwa8_info
--   12: stg_catch_frame_info
-- @
--
-- For a program looking like this
--
-- @
-- module Main where
--
-- import GHC.ExecutionStack
-- import System.IO.Unsafe (unsafePerformIO)
--
-- main :: IO()
-- main = print (dontInlineMe (show (unsafePerformIO printStack)))
--
-- printStack :: IO ()
-- printStack = currentStackFrames >>= (putStrLn . displayStackFrames)
--
-- dontInlineMe :: String -> String
-- dontInlineMe x = case x of
--                    "hello" -> "world"
--                    x       -> x
-- @
--
-- As can be seen, the stack trace reveals function names.
--
displayStackFrames :: [StackFrame] -> String
displayStackFrames frames =
    "Stack trace (Haskell):\n" ++
    concatMap (uncurry displayFrame) ([0..] `zip` frames)

-- | How one StackFrame is displayed in one stack trace
displayFrame :: Int -> StackFrame -> String
displayFrame ix frame = unlines $ zipWith ($) formatters strings
      where formatters = (printf "%4u: %s" (ix :: Int)) : repeat ("      " ++)
            strings    = prepareStackFrame frame

-- We use these three empty data declarations for type-safety
data CodemapUnit
data CodemapProc
data Instruction

peekCodemapUnitName :: Ptr CodemapUnit -> IO CString
peekCodemapUnitName ptr = #{peek struct CodemapUnit_, name } ptr

peekCodemapProcName :: Ptr CodemapProc -> IO CString
peekCodemapProcName ptr = #{peek struct CodemapProc_, name } ptr

currentStackFrames :: IO ([StackFrame])
currentStackFrames = currentExecutionStack >>= getStackFrames

-- | Reify the stack. This is the only way to get an ExecutionStack value.
currentExecutionStack :: IO (ExecutionStack)
currentExecutionStack = currentExecutionStackLimit (maxBound :: Int)

currentExecutionStackLimit :: Int -> IO (ExecutionStack)
currentExecutionStackLimit (I## i##) =
    IO (\s -> let (## new_s, byteArray## ##) = reifyStack## i## s
                  ba = ExecutionStack byteArray##
              in (## new_s, ba ##) )

-- | Tell the codemap module that you want to use codemap. Synchronized.
foreign import ccall "Codemap.h codemap_inc_ref" codemapIncRef :: IO ()

-- | Tell the codemap module that you are done using codemap. Synchronized.
foreign import ccall "Codemap.h codemap_dec_ref" codemapDecRef :: IO ()

-- | Ask the codemap module if it can unload and free up memory. It will not be
-- able to if the module is in use or if data is not loaded. Synchronized.
--
-- Returns True if codemap data was unloaded.
foreign import ccall "Codemap.h codemap_try_unload" codemapTryUnload :: IO Bool

foreign import ccall "Codemap.h codemap_is_loaded" codemapIsLoaded :: IO Bool

-- | Lookup an instruction pointer
--
-- Codemap module must be loaded to use this!
foreign import ccall "Codemap.h codemap_lookup_ip"
    codemapLookupIp ::
       Ptr Instruction -- ^ Code address you want information about
    -> Ptr (Ptr CodemapProc) -- ^ Out: CodemapProc Pointer Pointer
    -> Ptr (Ptr CodemapUnit) -- ^ Out: CodemapUnit Pointer Pointer
    -> IO ()

withCodemap :: IO a -> IO a
withCodemap = bracket_ codemapIncRef codemapDecRef

getStackFrameNoSync ::
       Ptr Instruction -- ^ Instruction Pointer
    -> IO StackFrame -- ^ Result
getStackFrameNoSync ip = do
    alloca $ \ppCodemapProc -> do
      poke ppCodemapProc nullPtr
      alloca $ \ppCodemapUnit -> do
          codemapLookupIp ip
                        ppCodemapProc
                        ppCodemapUnit
          pCodemapProc <- peek ppCodemapProc
          pCodemapUnit <- peek ppCodemapUnit
          unitName <- stringPeekWith peekCodemapUnitName pCodemapUnit
          procedureName <- stringPeekWith peekCodemapProcName pCodemapProc
          return StackFrame{..}

-- Note: if you grepped your way to the string "<Data not found>,
-- you probably forgot to compile that module with the `-g` flag to ghc.
stringPeekWith :: (Ptr a -> IO CString) -> Ptr a -> IO String
stringPeekWith _peeker ptr | ptr == nullPtr = return "<Data not found>"
stringPeekWith peeker ptr  | otherwise      = peeker ptr >>= peekCString

getStackFrames :: ExecutionStack -> IO [StackFrame]
getStackFrames stack = withCodemap $ getStackFramesNoSync stack

getStackFramesNoSync :: ExecutionStack -> IO [StackFrame]
getStackFramesNoSync = mapM getStackFrameNoSync . stackIndices

