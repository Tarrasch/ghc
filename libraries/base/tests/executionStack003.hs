import GHC.ExecutionStack
import Debug.Trace
import GHC.IO (unsafePerformIO)
import Data.List (isPrefixOf, replicate)

-- Are ticks across specializations maintained?
--
-- I also did eyeball the output of
--
-- $ ./inplace/bin/ghc-stage2 -ddump-simpl -fspecialise -O \
--     libraries/base/tests/executionStack003.hs | less
--
-- There I make sure that the general version gets replaced with one
-- specialized for Int.
--
-- Also, it might be interesting to pass the flags '-g -dppr-ticks' if you
-- want to do manual inspection

myTypeclassFunction :: (Num a) => [a] -> a
{-# SPECIALISE myTypeclassFunction :: [Int] -> Int #-}
myTypeclassFunction []     = fromIntegral $ unsafePerformIO $ findOccurencesOf "myTypeclassFunction"
myTypeclassFunction (x:xs) = myTypeclassFunction xs + x

findOccurencesOf :: String -> IO Int
findOccurencesOf fname = do
    stack <- currentExecutionStack
    stackFrames <- getStackFrames stack
    let infos = concatMap locationInfos stackFrames
        functionNames = map functionName infos
    return $ length . filter (isPrefixOf fname) $ functionNames

formatter x | x > 20    = "Wut?"
            | x > 2     = "Dwarf data remains when specializing! :)"
            | otherwise = "Debug data gets lost when specializing"

main = putStrLn
      . formatter
      . (myTypeclassFunction :: [Int] -> Int)
      $ replicate 5 (0 :: Int)
     -- XXX: We only recurse a few times in myTypeclassFunction,
     -- otherwise we'll get marked update frames, and we won't be able to
     -- resolve the function names on them.

-- Historical note: Back when I worked on a ghc branch from late 2013, before
-- 7822166, I also had a test where I tested that the sum function gets put on
-- the stack. After rebasing I suddenly saw that my test case broke because the
-- sum function is now defined as a fold and is inlined. :)
