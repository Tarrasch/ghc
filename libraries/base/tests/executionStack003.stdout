Dwarf data remains when specializing! :)
